File: Array_problems/RotateImage.cpp

// Problem link - https://leetcode.com/problems/rotate-image/description/?envType=daily-question&envId=2026-05-04

#include<bits/stdc++.h>
using namespace std;

void rotate(vector<vector<int>>& matrix) {
    int n = matrix.size();

    // step 1 - find transpose of matrix
    for(int i=0; i<n; i++){
        for(int j=i; j<n; j++){
            swap(matrix[i][j], matrix[j][i]);
        }
    }

    // step 2 - reverse rows of the matrix
    for(int i=0; i<n; i++){
        reverse(matrix[i].begin(), matrix[i].end());
    }
}

void print2DVector(vector<vector<int>>& matrix){
    cout << "[";
    for (size_t i = 0; i < matrix.size(); ++i) {
        cout << "[";
        for (size_t j = 0; j < matrix[i].size(); ++j) {
            cout << matrix[i][j];
            if (j < matrix[i].size() - 1) cout << ",";
        }
        cout << "]";
        if (i < matrix.size() - 1) cout << ",";
    }
    cout << "]" << std::endl;
}

int main(){
    // vector<vector<int>> matrix = {{1,2,3},{4,5,6},{7,8,9}};

    vector<vector<int>> matrix = {{5,1,9,11},{2,4,8,10},{13,3,6,7},{15,14,12,16}};
    rotate(matrix);
    print2DVector(matrix);
    return 0;
}

File: Array_problems/RotateImage.java

// Problem link - https://leetcode.com/problems/rotate-image/description/?envType=daily-question&envId=2026-05-04

package Array_problems;

public class RotateImage {
    public void rotate(int[][] matrix) {
        int n = matrix.length;

        // step 1 - find transpose of matrix
        for(int i=0; i<n; i++){
            for(int j=i; j<n; j++){
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }

        // step 2 - reverse each row of matrix
        for(int i=0; i<n; i++){
            int s = 0;
            int e = n - 1;

            while(s < e){
                int temp = matrix[i][s];
                matrix[i][s] = matrix[i][e];
                matrix[i][e] = temp;

                s++;
                e--;
            }
        }
    }

    public static void main(String[] args) {
        int[][] matrix = {{1,2,3},{4,5,6},{7,8,9}};
        RotateImage obj = new RotateImage();
        obj.rotate(matrix);

    }
}
